### ISSUE TITLE
Review for DL@MBL 2024

### ISSUE THREAD

**Issue by ziw-liu** [dlmbl/intro_dl#5]
The notebook works great! I have no major concerns, only small editing suggestions:

## Setup

- [x] We are using VS Code for this year's course. The README still calls for Jupyter Lab:

https://github.com/dlmbl/intro_dl/blob/362b625bd80d689ddf82f86c3a52db6c4ffb16fa/README.md?plain=1#L31-L36

- [x] Consider bumping Python to a version recommended by [SPEC-0](https://scientific-python.org/specs/spec-0000/).

## Task 1

- [x] Consider adding 'MLP' as the common abbreviation when introducing multi-layer perceptron:

https://github.com/dlmbl/intro_dl/blob/362b625bd80d689ddf82f86c3a52db6c4ffb16fa/solution.py#L18

- [x] Is the 'deep' CNN really deep? Maybe use 'deeper' instead.

https://github.com/dlmbl/intro_dl/blob/362b625bd80d689ddf82f86c3a52db6c4ffb16fa/solution.py#L19

- [x] The step function outputs a boolean, which is different from the text saying it should be $f: \mathbb{R}\mapsto\mathbb{R}$. It might also be worth noting that a step function will not be usable in gradient descent.

https://github.com/dlmbl/intro_dl/blob/362b625bd80d689ddf82f86c3a52db6c4ffb16fa/solution.py#L72-L73

- [x] The Markdown style does not render properly inside of an HTML block. Consider using an HTML tag (e.g. `<code>f</code>` -> <code>f</code>) instead.

https://github.com/dlmbl/intro_dl/blob/362b625bd80d689ddf82f86c3a52db6c4ffb16fa/solution.py#L119

## Task 2

- [x] Small grammar fixes needed in:

https://github.com/dlmbl/intro_dl/blob/362b625bd80d689ddf82f86c3a52db6c4ffb16fa/solution.py#L180

## Task 3

- [ ] How about adding a plot showing the loss curve during training like the one for CNN below?

- [ ] Consider using [`with torch.inference_mode(): ...`](https://pytorch.org/docs/stable/generated/torch.autograd.grad_mode.inference_mode.html).

https://github.com/dlmbl/intro_dl/blob/362b625bd80d689ddf82f86c3a52db6c4ffb16fa/solution.py#L453

- [x] Clarify why using a non-learnable downsampling layer is always better than a strided convolution. AFAIK this is not true for all learning tasks.

https://github.com/dlmbl/intro_dl/blob/362b625bd80d689ddf82f86c3a52db6c4ffb16fa/solution.py#L703

## Task 4

- [x] Why does the download path needs to be hidden?

https://github.com/dlmbl/intro_dl/blob/362b625bd80d689ddf82f86c3a52db6c4ffb16fa/solution.py#L717-L722

**Comment by AlbertDominguez** [comment]
Thx for the feedback Ziwen! I've assessed most of it already. Regarding two points

> How about adding a plot showing the loss curve during training like the one for CNN below?

I thought of doing that, but i wanted to discuss training curves, etc., at the end of the exercise, and having them without giving some concepts felt a bit strange. Might add it though, maybe we can discuss on prep day.

> Consider using [with torch.inference_mode()](https://pytorch.org/docs/stable/generated/torch.autograd.grad_mode.inference_mode.html

I'm already using it! Are you meaning that the `.eval()` call can be removed inside that context manager? In any case, I think it's good to have the `.eval()` explicitly anyway to avoid the classic "Ouch I missed the `.eval()`", in case someone uses/reads `torch.no_grad()`, better safe than sorry!

**Comment by ziw-liu** [comment]
> Are you meaning that the .eval() call can be removed inside that context manager?

Yes. I think `torch.inference_mode()` is now recommended over `nn.Module.eval()` since it is a superset of the later with additional inference time behavior. In this case it's probably worth pointing out in the comment that you actually don't need both.

**Comment by ziw-liu** [comment]
There is a typo in the updated version:
https://github.com/dlmbl/intro_dl/blob/f8a9e8ee9a5ad45373b8d9933a2d551e35beff0d/solution.py#L75
